### PR TITLE
Feat check certificate validity

### PR DESCRIPTION
**Pull request by vermakhushboo**
## What does this PR do?

Get status of certificate for a domain to check whether it is still valid or expired.

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Screenshots may also be helpful.)

## Related PRs and Issues

- (Related PR or issue)

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: app/controllers/api/health.php [hunk 1]
```diff
@@ -388,6 +389,33 @@
...
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'health')
+    ->label('sdk.method', 'getCertificate')
+    ->label('sdk.description', '/docs/references/health/get-certificate.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_HEALTH_STATUS)
+    ->param('domain', null, new Domain(), 'Domain name')
+    ->inject('response')
+    ->action(function (string $domain, Response $response) {
```
Comment (Meldiron) [inline app/controllers/api/health.php:405]: Let's add a permission check here.
If I provide API key of my project which has a custom domain added, then the health check should run. but if I try to validate domain which doesn't belong to my project, it should give 404

File: app/controllers/api/health.php [hunk 2]
```diff
@@ -388,6 +389,33 @@
...
     }, ['response']);
 
+App::get('/v1/health/certificate')
+    ->desc('Get status of certificate for a domain to check whether it is still valid or expired.')
+    ->groups(['api', 'health'])
+    ->label('scope', 'health.read')
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'health')
+    ->label('sdk.method', 'getCertificate')
+    ->label('sdk.description', '/docs/references/health/get-certificate.md')
```
Comment (Meldiron) [inline app/controllers/api/health.php:399]: Markdown file missing

File: app/controllers/api/health.php [hunk 3]
```diff
@@ -388,6 +391,53 @@
...
+    ->inject('response')
+    ->action(function (string $domain, Response $response) {
+        if (filter_var($domain, FILTER_VALIDATE_URL)) {
+            $domain = parse_url($domain, PHP_URL_HOST);
+        }
+
+        $get = stream_context_create(array("ssl" => array("capture_peer_cert" => true)));
+        $read = stream_socket_client("ssl://" . $domain . ":443", $errno, $errstr, 30, STREAM_CLIENT_CONNECT, $get);
+        if (!$read) {
+            throw new Exception(Exception::INVALID_HOST, 'The domain is not valid.');
```
Comment (christyjacob4) [inline app/controllers/api/health.php:415]: You dont need to pass an error code and a description. Just set the error code and the description is automatically picked up from the errors.php file. Only if we need to override the base description, we need to pass it here.

File: src/Appwrite/Utopia/Response/Model/HealthCertificate.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                    'signatureTypeSN' => 'RSA-SHA256',
+                ],
+            ])
+            ->addRule('status', [
+                'type' => self::TYPE_STRING,
+                'description' => 'Service status. Possible values can are: `pass`, `fail`',
+                'default' => '',
+                'example' => 'pass',
+            ])
+        ;
```
Comment (christyjacob4) [inline src/Appwrite/Utopia/Response/Model/HealthCertificate.php:36]: Why do we need a status? The HTTP status code should reflect the status of the certificate. 200 for success and a relevant code for the failed certificates

File: app/config/errors.php [hunk 5]
```diff
@@ -795,6 +795,18 @@
         'publish' => false
     ],
 
+    Exception::CERTIFICATE_EXPIRED => [
+        'name' => Exception::CERTIFICATE_EXPIRED,
+        'description' => 'Certificate expired.',
```
Comment (christyjacob4) [inline app/config/errors.php:800]: We need to have a more verbose description here

File: app/config/errors.php [hunk 6]
```diff
@@ -795,6 +795,18 @@
...
+        'code' => 404,
+    ],
+
+    Exception::INVALID_HOST => [
+        'name' => Exception::INVALID_HOST,
+        'description' => 'Host not found.',
```
Comment (christyjacob4) [inline app/config/errors.php:806]: same as above

File: app/controllers/api/health.php [hunk 7]
```diff
@@ -388,6 +391,53 @@
...
+    ->label('sdk.response.model', Response::MODEL_HEALTH_CERTIFICATE)
+    ->param('domain', null, new Multiple([new Domain(), new PublicDomain()]), Multiple::TYPE_STRING, 'Domain name')
+    ->inject('response')
+    ->action(function (string $domain, Response $response) {
+        if (filter_var($domain, FILTER_VALIDATE_URL)) {
+            $domain = parse_url($domain, PHP_URL_HOST);
+        }
+
+        $get = stream_context_create(array("ssl" => array("capture_peer_cert" => true)));
+        $read = stream_socket_client("ssl://" . $domain . ":443", $errno, $errstr, 30, STREAM_CLIENT_CONNECT, $get);
```
Comment (christyjacob4) [inline app/controllers/api/health.php:413]: similar comment about the variable name

File: app/controllers/api/health.php [hunk 8]
```diff
@@ -388,6 +391,53 @@
...
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_HEALTH_CERTIFICATE)
+    ->param('domain', null, new Multiple([new Domain(), new PublicDomain()]), Multiple::TYPE_STRING, 'Domain name')
+    ->inject('response')
+    ->action(function (string $domain, Response $response) {
+        if (filter_var($domain, FILTER_VALIDATE_URL)) {
+            $domain = parse_url($domain, PHP_URL_HOST);
+        }
+
+        $get = stream_context_create(array("ssl" => array("capture_peer_cert" => true)));
```
Comment (christyjacob4) [inline app/controllers/api/health.php:412]:
```suggestion
        $get = stream_context_create(["ssl" => ["capture_peer_cert" => true]]);
```
In our code base, we don't really use `array()` to create new arrays. we use the array literal. Also please format it to multiple lines

File: app/controllers/api/health.php [hunk 9]
```diff
@@ -388,6 +391,53 @@
         $response->dynamic(new Document([ 'size' => $size ]), Response::MODEL_HEALTH_QUEUE);
     }, ['response']);
 
+App::get('/v1/health/certificate')
+    ->desc('Get status of certificate for a domain to check whether it is still valid or expired.')
```
Comment (christyjacob4) [inline app/controllers/api/health.php:395]:
```suggestion
    ->desc('Get the SSL certificate for the domain')
```

File: app/config/errors.php [hunk 10]
```diff
@@ -795,6 +795,18 @@
         'publish' => false
     ],
 
+    Exception::CERTIFICATE_EXPIRED => [
```
Comment (christyjacob4) [inline app/config/errors.php:798]: Let's make sure the health error codes have a `HEALTH_` prefix

**Review by eldadfux — APPROVED** [review]

**Comment by vermakhushboo** [comment]
We've already merged this to cloud: https://github.com/appwrite/appwrite/pull/7547
Do we close this PR?

**Pull request closed without merging**